spec/deploy_spec.rb

```diff
@@ -60,6 +64,70 @@
           expect(@user_comms).to have_received(:say_thank_you)
         end
       end
+
+      context 'when there is already a develop tag on the commit'do
+        before(:each) do
+          allow(@test_and_stage_tag_generator).to receive(:check_for_tag?).and_return(true)
+        end
+
+        it "tells the user that there is already at dev tag on the commit" do
+          @deploy.environment_choice
+          expect(@user_comms).to have_received(:error_commit_has_dev_tag)
+        end
+      end
+
+      context 'when there are no tags on the repository'do
+        before(:each) do
+          allow(@test_and_stage_tag_generator).to receive(:check_for_tag?).and_return(false)
+          allow(@develop_tag_generator).to receive(:develop_tag_exists?).and_return(false)
+
+          allow(@user_comms).to receive(:tell_user_no_develop_tags)
+          allow(@develop_tag_generator).to receive(:first_tag).and_return('dev-v0.0.1')
+
+          allow(@user_comms).to receive(:ask_permissison_to_apply).with('dev-v0.0.1')
+          allow(@user_comms).to receive(:user_reply_y_or_n).and_return('y')
+        end
+
+        it "applies the first tag" do
+          @deploy.environment_choice
+          expect(@git_helper).to have_received(:apply_tag).with('dev-v0.0.1')
+        end
+
+        it "pushes the first tag" do
+          @deploy.environment_choice
+          expect(@git_helper).to have_received(:push_tag_to_remote).with('dev-v0.0.1')
+        end
+      end
     end
+
+    describe 'test' do
+      context ' when the user would like to apply a test tag'do
+      before (:each) do
+        @deploy = DeployController.new('test', @git_helper, @user_comms, @develop_tag_generator, @test_and_stage_tag_generator)
```

Comment:
We also usually use the `subject{}` notation to signify what specific part is going to be tested.

`subject{DeployController.new('test', @git_helper, @user_comms, @develop_tag_generator, @test_and_stage_tag_generator)}`

and then

`subject.environment_choice` etc.